test/distributed/fsdp/test_fsdp_misc.py

```diff
@@ -143,6 +143,109 @@ def _check_device_matches(module, device_id):
             nested_wrapped_module, torch.device("cuda", torch.cuda.current_device())
         )
 
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp_zero2_eval_with_prefetch(self):
+        # Test FSDP validation with SHARD_GRAD_OP and forward_prefetch
+
+        class Mnist(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.conv1 = nn.Conv2d(1, 32, 3, 1)
+                self.conv2 = nn.Conv2d(32, 64, 3, 1)
+                self.dropout1 = nn.Dropout(0.25)
+                self.dropout2 = nn.Dropout(0.5)
+                self.fc1 = nn.Linear(9216, 128)
+                self.fc2 = nn.Linear(128, 10)
+                self.ln = nn.LayerNorm(9216)
+
+            def forward(self, x, y):
+                x = self.conv1(x)
+                x = torch.nn.functional.relu(x)
+                x = self.conv2(x)
+                x = torch.nn.functional.relu(x)
+                x = torch.nn.functional.max_pool2d(x, 2)
+                x = self.dropout1(x)
+                x = torch.flatten(x, 1)
+                x = self.ln(x)
+                x = self.fc1(x)
+                x = torch.nn.functional.relu(x)
+                x = self.dropout2(x)
+                x = self.fc2(x)
+                output = torch.nn.functional.log_softmax(x, dim=1)
+                loss = torch.nn.functional.cross_entropy(output, y)
+                return loss
+
+        model = Mnist().cuda()
+        model1 = Mnist().cuda()
+        model1.load_state_dict(model.state_dict())
+        fsdp_model = FSDP(
+            model,
+            sharding_strategy=ShardingStrategy.SHARD_GRAD_OP,
+            forward_prefetch=True,
+            use_orig_params=True,
+            auto_wrap_policy=ModuleWrapPolicy([nn.Linear, nn.Conv2d]),
+        )
+        ddp_model = torch.nn.parallel.DistributedDataParallel(
+            model1,
+        )
+
+        fsdp_opt = torch.optim.SGD(fsdp_model.parameters(), lr=1e-4)
+        ddp_opt = torch.optim.SGD(ddp_model.parameters(), lr=1e-4)
+
+        x = torch.randn(8, 1, 28, 28, device="cuda").requires_grad_()
+        y = torch.randint(low=0, high=9, size=(8,), device="cuda")
+        x1 = x.clone().detach().requires_grad_()
+        y1 = y.clone().detach()
+
+        seed = self.rank + 20231010
+
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed(seed)
+        for _ in range(5):
+            fsdp_loss = fsdp_model(x, y)
+            fsdp_loss = fsdp_loss.sum()
+            fsdp_loss.backward()
+            fsdp_opt.step()
+
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed(seed)
+        for _ in range(5):
+            ddp_loss = ddp_model(x1, y1)
+            ddp_loss = ddp_loss.sum()
+            ddp_loss.backward()
+            ddp_opt.step()
+
+        assert torch.allclose(fsdp_loss, ddp_loss)
+        assert torch.allclose(x.grad, x1.grad)
```

Comment:
nit: Would it be possible to check correctness after every iteration? This lets us narrow down / catch bugs much faster.

You can reference: https://github.com/pytorch/pytorch/pull/110948/files#diff-ab5af580410c642dd66ea27656265fbbc1ec6c6713e048a0ef111573eb52286cR195

Reply:
I've perfected this UT, and by the way, the author of the PR above turned out to be a coworker of mine😂.